macros/edr/dbt_artifacts/upload_artifacts_to_table.sql

```diff
@@ -18,4 +18,61 @@
     {%- if should_commit -%}
         {% do adapter.commit() %}
     {%- endif -%}
-{% endmacro %}
\ No newline at end of file
+{% endmacro %}
+
+{% macro upload_csv_artifacts_to_table(table_relation, artifacts, flatten_artifact_callback) %}
+    {% set output_csv_path = elementary.get_target_path(table_relation.identifier ~ '.csv') %}
+    {% set flatten_artifact_dicts = [] %}
+    {% for artifact in artifacts %}
+        {% set flatten_artifact_dict = flatten_artifact_callback(artifact) %}
+        {% set flatten_artifact_rendered_dict = {} %}
+        {% for key in flatten_artifact_dict %}
+          {% set value = flatten_artifact_dict[key] %}
+          {% if value %}
+            {% do flatten_artifact_rendered_dict.update({key: elementary.render_csv_value(value)}) %}
+          {% endif %}
+        {% endfor %}
+        {% do flatten_artifact_dicts.append(flatten_artifact_rendered_dict) %}
+    {% endfor %}
+    {% set flatten_artifacts_agate = elementary.get_agate_table().from_object(flatten_artifact_dicts) %}
+    {% set cached_artifacts_agate = elementary.get_cached_artifacts_agate(table_relation, output_csv_path) %}
+    {% if elementary.are_artifacts_equal(cached_artifacts_agate, flatten_artifacts_agate) %}
+      {% do elementary.debug_log("[%s] Artifacts were not changed. Skipping upload." % table_relation.identifier) %}
+      {% do return(none) %}
+    {% endif %}
+    {% do elementary.debug_log("[%s] Uploading artifacts." % table_relation.identifier) %}
+    {% do elementary.seed_elementary_model(table_relation, output_csv_path) %}
+    {% do flatten_artifacts_agate.to_csv(output_csv_path) %}
+{% endmacro %}
+
+{% macro get_cached_artifacts_agate(table_relation, cache_csv_path) %}
+  {% if cache_csv_path.exists() %}
```

Comment:
Not sure that this macro should be responsible for saving to disk the returned artifacts from the db.. wdyt about returning here the cached artifacts and saving it to local disk only if you need to upload (meaning there's a change)

Reply:
Interestingly, that was what I originally did.
The thing is that file-caching is much faster than db-caching,
if an environments runs those models but it already has artifacts (say a new engineer joined),
he'll never get the file-caching until he'll change something in the project.

Reply:
Got it but I'd still write it outside of this macro (like you can move line 45 to be before the `are_artifacts_equal` validation

Reply:
That is also a problem because this could case a discrepancy if the upload failed.
You'd overwrite the cache before you know you successfully made the update.